### PR TITLE
Add Kubernetes Cluster Connection Tester

### LINKED ISSUE #16824: Discover: Provide ability to test Kubernetes cluster connection
https://github.com/gravitational/teleport/pull/15413 implemented API for connection diagnostics/tester for SSH node connections used as part of the new Teleport Discover flow. We need to extend this API to support testing connections to Kubernetes clusters.

### PR DESCRIPTION
**Pull request by tigrato**
This PR implements the Kubernetes Tester for the ConnectionDiagnostic feature.

The goal is to validate if the Kubernetes cluster is available and if the user can access it.

A series of checks are performed against Teleport using a Kubernetes Client populated with the user's certificates for accessing the desired cluster.

Currently, the checks are:

- Kubernetes Cluster is up and running
- User's role specifies at least one `kubernetes_groups` or `kubernetes_users`
- User's roles allow RBAC access to the cluster: `kubernetes_labels` match.
- Values specified under `kubernetes_groups` or `kubernetes_users` exist in the cluster and allow the user to list pods in the desired namespace.

Testing:
Post: `/v1/webapi/sites/<teleport-cluster>/diagnostics/connections`

```json
{
  "resource_kind": "kube_cluster",
  "resource_name": "<cluster_name>",
   "kubernetes_namespace": "default"
}

```

Closes #16824

### CODE REVIEW

**Comment by tigrato** [comment]
cc @kimlisa @ryanclark

File: lib/web/apiserver.go
Comment (ibeckermayer) [inline lib/web/apiserver.go:3091]: I think you can abstract out the common logic between this and `ProxyHostPort` like

```go
// ProxyHostPort returns the address of the proxy server using --proxy
// notation, i.e. "localhost:8030,8023"
func (h *Handler) ProxyHostPort() string {
	hp := createHostPort(h.cfg.ProxyWebAddr, defaults.HTTPListenPort)
	return fmt.Sprintf("%s,%s", hp, h.sshPort)
}

func (h *Handler) kubeProxyHostPort() string {
	return createHostPort(h.cfg.ProxyKubeAddr, defaults.KubeListenPort)
}

// Address can be set in the config with unspecified host, like
// 0.0.0.0:3080 or :443.
//
// In this case, the dial will succeed (dialing 0.0.0.0 is same a dialing
// localhost in Go) but the host certificate will not validate since
// 0.0.0.0 is never a valid principal (auth server explicitly removes it
// when issuing host certs).
//
// As such, replace 0.0.0.0 with localhost in this case: proxy listens on
// all interfaces and localhost is always included in the valid principal
// set.
func createHostPort(netAddr utils.NetAddr, port int) string {
	if netAddr.IsHostUnspecified() {
		return fmt.Sprintf("localhost:%v", netAddr.Port(port))
	}
	return netAddr.String()
}
```

File: lib/client/conntest/kube.go
Comment (tigrato) [reply lib/client/conntest/kube.go:198]: Except for the last [test](https://github.com/gravitational/teleport/pull/16899/files/e3104515faf1ead75403db140c0e3103e7c24075#diff-b2ceb7548b64fc48f2d85f2f90af15824433e53a610508be9d8785325feb6335R221), all strings are generated by the Teleport Kube Service.
The last test, K8S RBAC, is a constant message since the RBAC was introduced in the Kubernetes repository, so I don't think we should worry about it just yet. When we start doing integration testing with real Kubernetes clusters, we can add support for the last message.

File: lib/client/conntest/kube.go
Comment (ryanclark) [inline lib/client/conntest/kube.go:254]: Should this happen here? It seems strange to me to take an `error` as a parameter and always return that error if it exists - I think this should be `handleUserGenCerts` and the check on `actionErr` should happen before getting to this method.

File: lib/client/conntest/kube.go
Comment (ryanclark) [inline lib/client/conntest/kube.go:214]: To not have such long `if` statements that require you to read the entire thing to understand what is being checked, I'd do something more like

```go
if kubeErr != nil {
	isNotFound := strings.Contains(kubeErr.ErrStatus.Message, "not found")
	isAccessDenied := strings.Contains(kubeErr.ErrStatus.Message, "[00] access denied")

	if isNotFound || isAccessDenied {
		message := "You are not authorized to access this Kubernetes Cluster. Ensure your role grants access by adding it to the 'kubernetes_labels' property."
		traceType := types.ConnectionDiagnosticTrace_RBAC_KUBE
		connDiag, err := s.appendDiagnosticTrace(ctx, connectionDiagnosticID, traceType, message, actionErr)
		return connDiag, trace.Wrap(err)
	}

	cannotListPods := strings.Contains(kubeErr.ErrStatus.Message, "cannot list resource \"pods\"")
	if cannotListPods {
		message := fmt.Sprintf("You are not allowed to list pods in the %q namespace. "+
			"Make sure your \"kubernetes_groups\" or \"kubernetes_users\" exist in the cluster and grant you access to list pods.", namespace)
		traceType := types.ConnectionDiagnosticTrace_KUBE_PRINCIPAL
		connDiag, err := s.appendDiagnosticTrace(ctx, connectionDiagnosticID, traceType, message, actionErr)
		return connDiag, trace.Wrap(err)
	}

	// etc...
}
```

Thoughts?

File: lib/client/conntest/kube.go
Comment (ryanclark) [reply lib/client/conntest/kube.go:198]: I did a global code search and can't see an error message that contains "has no assigned groups or users" coming anywhere other than Teleport. Is it possible for k8s to return this error?

https://cs.github.com/?q=%22has+no+assigned+groups+or+users%22&p=1&pt=b90841e231f5ad440693d6711568e5ed53e05ad9bb1d125abdb65081e2ae597260ee68a103e70d0756ab3b28caf80227d7f197de8b5725f8373434a82427ada1dad4efa53929a3b3c1e37b63855e048d9f6276b9eda4222e8a2b85688804d42c3e850399274d322dc3fa89eaf1e1e789d64b9803251d25b95375e739bc1768535304f3ed3ed49cc2&scope=&scopeName=All+repos

Comment (tigrato) [reply lib/client/conntest/kube.go:198]: No, this message is generated from Teleport only. Kubernetes does not have the idea of assigned user/groups. If you have no users or groups, you are automatically converted to `system:anonymous` and the request continues.

File: lib/client/conntest/kube.go
Comment (ryanclark) [inline lib/client/conntest/kube.go:252]: nit pick 
```suggestion
	}
	
	if actionErr != nil {
```

You're returning in your first branch so no need to do an `else`

File: lib/client/conntest/kube.go
Comment (ryanclark) [reply lib/client/conntest/kube.go:198]: Ah that makes more sense. Could we have these errors hard-coded as constants, to remove the possibility of drift?

File: lib/client/conntest/kube.go
Comment (tigrato) [reply lib/client/conntest/kube.go:183]: This is related to `SSH` connection testing mechanism. 

Since the SSH conn testing offloads the execution to `tsh ssh ....` we are required to init the testing and embed the `connectionDiagnosticID` inside the SSH Certificate extension. Teleport SSH service reads the `connectionDiagnosticID` from the certificate and appends the desired data and steps into the disk so we can extract it at the ssh connection testing once it finishes.

I followed the same global idea although here we create a K8S client directly to kubeproxy.

Comment (ibeckermayer) [reply lib/client/conntest/kube.go:185]: I see, in that case I agree that consistent behavior between the two diagnostics is best.

File: lib/client/conntest/kube.go
Comment (tigrato) [reply lib/client/conntest/kube.go:200]: @ryanclark @ibeckermayer we can't. Those messages include user data (roles, groups, name...) that changes but they are covered by unit tests using the real components.

Comment (ibeckermayer) [reply lib/client/conntest/kube.go:200]: It's not true that we can't, for example it looks like comes from "has no assigned groups or users"
https://github.com/gravitational/teleport/blob/504db74e8b59c6f48b074c4773b289e8061f9e36/lib/services/role.go#L1304 which can be pulled out into a const.

"[00] access denied" comes from 
https://github.com/gravitational/teleport/blob/504db74e8b59c6f48b074c4773b289e8061f9e36/lib/kube/proxy/forwarder.go#L398
which can just be pulled out of that function (and exported if necessary).

If e.g. "not found" is part of some longer message, it can be extracted out and then placed in the larger message via `fmt.Sprintf("something something %s, NotFoundErrStr)`.

File: lib/web/apiserver_test.go
Comment (nklaassen) [inline lib/web/apiserver_test.go:5543]: I don't see a purpose to `closeC`

File: lib/service/service.go
Comment (r0mant) [inline lib/service/service.go:3525]: Should this be listen address or public address?

File: lib/client/conntest/kube.go
Comment (r0mant) [inline lib/client/conntest/kube.go:111]: Can this logic be cleaned up a bit, for example I would consider factoring out parts of this function into dedicated method e.g . "createKubeClient", etc.

File: lib/client/conntest/kube.go
Comment (r0mant) [inline lib/client/conntest/kube.go:158]: This is a little confusing tbh. Is this required to avoid x509 errors in case proxy's public address is `localhost:xxx`? Shouldn't this logic be part of GetKubeTLSServerName then?

File: lib/client/conntest/kube.go
Comment (r0mant) [inline lib/client/conntest/kube.go:255]:
```suggestion
		message := "Failed to connect to Kubernetes cluster. Ensure the cluster is registered."
```

**Review by r0mant — APPROVED** [review]
lgtm once remaining feedback is resolved

File: lib/web/apiserver_test.go
Comment (tigrato) [reply lib/web/apiserver_test.go:5541]: Didn't know that!

**Review by ryanclark — APPROVED** [review]

File: lib/client/conntest/kube.go
Comment (ibeckermayer) [reply lib/client/conntest/kube.go]: Good point, I hadn't recognized that the source of the messages could follow a separate update cycle. In that case it's perhaps worth adding a comment near the hardcode messages or the tests noting this subtlety, in case it does become an issue, (your call whether you think that's worthwhile or not). Approved.

**Review by ibeckermayer — APPROVED** [review]

**Pull request merged**
